### PR TITLE
VACMS-4628: Add revision log message validation to vcs.

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
See #4628

## Testing done
Visual

## QA steps
As an administrator, edit any existing vc node type without adding a revision log message. 
Visually verify that a form error is thrown, indication you need to add a revision log message.
Add a message, and confirm that form saves without error being thrown.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Core Application Team`
- [x] `Product Support Team`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change. 
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 1]
```diff
@@ -415,6 +415,18 @@ function _va_gov_backend_allowed_formats_remove_textarea_help(array $form_elemen
...
+    'node_vet_center_cap_edit_form',
+    'node_vet_center_facility_health_servi_edit_form',
+    'node_vet_center_locations_list_edit_form',
+    'node_vet_center_mobile_vet_center_edit_form',
+    'node_vet_center_outstation_edit_form',
+  ];
+  // Vet centers need to have revision log messages on edit.
+  if (in_array($form_id, $vc_types)) {
+    $form['#validate'][] = '_va_gov_backend_require_revision_message';
+  }
```
[the remark this answers is not in the document]
Comment (kevwalsh) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:429]: Not pushing back on this going to va_gov_vet_center, but i think it's likely this becomes more widely adopted by other content types (possibly all!).

[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:429]: @ethanteague to Kevs point  ^^ ok to leave it in va_gov_backend, lets just move it into a call to
_va_gov_backend_require_revision_message($form, $form_id)

[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:429]: Updated, @swirtSJW @kevwalsh

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 2]
```diff
@@ -568,6 +580,21 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
...
+ * Adds Validation to check for an empty Appointment lead-in field.
+ *
+ * @param array $form
+ *   The exposed widget form array.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _va_gov_backend_require_revision_message(array $form, FormStateInterface &$form_state) {
+  // Add revision log validation.
+  if (empty($form_state->getValue('revision_log')[0]['value'])) {
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:593]: I don't think we can avoid it, here, @swirtSJW. Those methods return the inverse of what is expected.
image

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 3]
```diff
@@ -568,6 +570,46 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
...
+    'node_vet_center_outstation_edit_form',
+  ];
+  // Vet centers need to have revision log messages on edit.
+  if (in_array($form_id, $vc_types)) {
+    $form['#validate'][] = '_va_gov_backend_validate_required_revision_message';
+  }
+}
+
+/**
+ * Adds Validation to check for an empty Appointment lead-in field.
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:599]: This seems to be copy pasta.

**Review by swirtSJW — CHANGES_REQUESTED** [review]

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module (hunk 2 above)
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:608]: `$form_state->isValueEmpty('revision_log')`  returns TRUE  of there is something in it?

[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:608]: Yep!

**Review by kevwalsh — CHANGES_REQUESTED** [review]
Forms seem to be invalidating properly, but i don't see a red asterisk to indicate required state. I feel like i remember them being there last time i QAd this.

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module (hunk 2 above)
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module]: I just update to this  `if ($form_state->isValueEmpty(['revision_log', '0', 'value'])) {`  It is still digging, but using the api to do it.


I also adjusted the error level so the link to the form element would work.

$form_state->setErrorByName('revision_log][0][value'

**Review by swirtSJW — APPROVED** [review]
Tested with single and bulk operations.  Works as intented.

**Comment by swirtSJW** [comment]
> Forms seem to be invalidating properly, but i don't see a red asterisk to indicate required state. I feel like i remember them being there last time i QAd this.

The link was pointing to an id that does not exist.  I resolved it.

**Comment by swirtSJW** [comment]
I misread Kev's comment   I fixed the validation error of the link not taking you to the form field, but yes the field is missing its required indicator.   Since it is not really a configurable field it is not just a quick config setting.  
this approach could work  but I hate scattering this logic all over the place..... though maybe just setting it required in teh base field would remove all the other custom validation... it would just work
https://www.drupal.org/project/drupal/issues/308352#comment-13371248

**Comment by ethanteague** [comment]
> Forms seem to be invalidating properly, but i don't see a red asterisk to indicate required state. I feel like i remember them being there last time i QAd this.

They weren't there, but I'll look for a way to implement this, @kevwalsh

**Comment by ethanteague** [comment]
Revision log required indicator added, @kevwalsh

**Comment by kevwalsh** [comment]
Hmm, no tugboat environment yet.

**Review by swirtSJW — APPROVED** [review]

**Comment by swirtSJW** [comment]
This tests out good 
image

**Comment by swirtSJW** [comment]
Looks like we have a failing test.

**Comment by ethanteague** [comment]
> Looks like we have a failing test.

Passes for me locally, @swirtSJW - just looped in @ElijahLynn in slack.

**Comment by ethanteague** [comment]
Tests passing, @swirtSJW

**Comment by ElijahLynn** [comment]
```
 root@php:/var/lib/tugboat# curl https://pr4911-iqbvykov5ruybyetpszrgwwzmm17z2ay.ci.cms.va.gov
The website encountered an unexpected error. Please try again later.<br /><em class="placeholder">Error</em>: Call to undefined method Drupal\user\Form\UserLoginForm::getEntity() in <em class="placeholder">va_gov_backend_form_alter()</em> (line <em class="placeholder">440</em> of <em class="placeholder">modules/custom/va_gov_backend/va_gov_backend.module</em>).
```

**Review by swirtSJW — APPROVED** [review]

**Comment by swirtSJW** [comment]
image

**Pull request merged**